### ISSUE TITLE
Possible to disable `editor.formatOnPaste` permanently?

### ISSUE THREAD

**Issue by prayash** [Tom-Bonnike/vscode-formatting-toggle#32]
I've never had success with the `editor.formatOnPaste` setting so I have that set to false but this plugin overrides it whenever it is toggled. I'm mostly interested in toggling `formatOnSave`.

Any thoughts on adding a flag to disable that? I can take a crack at it and open a PR 😄

**Comment by Tom-Bonnike** [comment]
https://github.com/Tom-Bonnike/vscode-formatting-toggle/blob/62a25bb78613838cd43d83cc728e7e46e8b90e9b/src/initCommand/index.ts#L39-L51
Haha “**probably** safe”.

In https://github.com/Tom-Bonnike/vscode-formatting-toggle/blob/master/src/initCommand.ts, I’m saving the initial `editor.…` configuration. This is only done once, when VSCode starts. This is useful so that I can set the value back to what it was for `formatOnType`, since I’m assuming very few people actually turn this setting on and so they wouldn’t want it put back to `true`.

I didn’t do it for `formatOnPaste` and `formatOnSave` because:
1) If I put those settings back to their initial value, they could both be set back to `false`… But as a user I would expect the extension to always activate them, even if they were deactivated when I first launched VS Code.
2) I simply assumed that’s what most people use and they shouldn’t cause any issue.

Could you please explain why you dislike `formatOnPaste`? 🤔Do you have any suggestion on how I could handle that? I can’t really guess if the user wants `formatOnPaste` or not… :/

---

Maybe I could add some settings that I would read? Could that be overkill? 😄 
`formattingToggle.activate = ['formatOnPaste', 'formatOnSave']`?

**Comment by prayasht** [comment]
@Tom-Bonnike Thanks for the prompt response! The reason I bring it up is because I've never found `formatOnPaste` to be desirable. In most cases it does more harm than good. Check out the demo below:

Notice how as I start to add a `newProp` to my React component, I paste the name of the method that I want to assign for the `newProp`'s value. Immediately on paste, it tries to format (in my case Prettier), which leads to a lot of visual confusion as the cursor has now moved and all the formatting of the file has gone haywire simply because I pasted in a method. This is one of many situations I've run into. Now imagine if `formatOnPaste` is off, I would paste, and finish making my changes, and save file. At that point my formatter would kick in and make whatever changes necessary, and I see a ✅ at the bottom right. I know I've done it right and move on.

![formatonpaste](https://user-images.githubusercontent.com/3971101/41195946-dd3785da-6bf3-11e8-9aea-e686a3cff1f5.gif)

Another reason is that I rely on my formatter to correct my syntax. Say if I've missed a `(` somewhere, my formatter would immediately show a `X` at the bottom corner of VS Code. So if my code doesn't format itself, I know I've missed something somewhere. It's a nice sanity check while saving changes onto a file, and I'd rather have that check happen on Save, and not on Paste. Because a paste doesn't mean that I'm done making my changes and now want it to be formatted, but a Save does.

I think your last suggestion is great. What if we ship some default options for what most users would use, and then allow the user to override? That way current users can keep using the Plugin as normal, and picky users like me can add an extra key to their settings JSON? 

Like you said, if I was able to add to my `settings.json` file a `formattingToggle.activate = ["formatOnSave"]`, my Formatting toggle would only formatOnSave, and I can toggle away and live happily ever after. 😄
